solidity/contracts/converter/types/liquidity-pool-v1/LiquidityPoolV1Converter.sol

```diff
@@ -177,27 +180,31 @@ contract LiquidityPoolV1Converter is LiquidityPoolConverter {
         return amount;
     }
 
-    function _write() internal {
-        uint256 price0 =
-            ISovrynSwapFormula(addressOf(SOVRYNSWAP_FORMULA)).crossReserveTargetAmount(
-                reserveBalance(reserveTokens[0]),
-                reserves[reserveTokens[0]].weight,
-                reserveBalance(reserveTokens[1]),
-                reserves[reserveTokens[1]].weight,
-                1
-            );
-
-        uint256 price1 =
-            ISovrynSwapFormula(addressOf(SOVRYNSWAP_FORMULA)).crossReserveTargetAmount(
-                reserveBalance(reserveTokens[1]),
-                reserves[reserveTokens[1]].weight,
-                reserveBalance(reserveTokens[0]),
-                reserves[reserveTokens[0]].weight,
-                1
-            );
-
-        oracle.write(price0, price1);
-    }
+	function _write() internal {
+		uint256 price0 = ISovrynSwapFormula(addressOf(SOVRYNSWAP_FORMULA)).crossReserveTargetAmount(
+			reserveBalance(reserveTokens[0]),
+			reserves[reserveTokens[0]].weight,
+			reserveBalance(reserveTokens[1]),
+			reserves[reserveTokens[1]].weight,
+			reserveBalance(reserveTokens[0]).div(POOL_PERCENTAGE)
+		);
+
+		uint256 price1 = ISovrynSwapFormula(addressOf(SOVRYNSWAP_FORMULA)).crossReserveTargetAmount(
+			reserveBalance(reserveTokens[1]),
+			reserves[reserveTokens[1]].weight,
+			reserveBalance(reserveTokens[0]),
+			reserves[reserveTokens[0]].weight,
+			reserveBalance(reserveTokens[1]).div(POOL_PERCENTAGE)
+		);
+
+		uint256 decimals0 = reserveTokens[0] == address(etherToken) ? ETHER_DECIMALS : IERC20Token(reserveTokens[0]).decimals();
+		uint256 decimals1 = reserveTokens[1] == address(etherToken) ? ETHER_DECIMALS : IERC20Token(reserveTokens[1]).decimals();

```

Comment:
having external calls for these each time this function is executed adds unnecessary gas cost. the reserve tokens do not change. the decimals could be stored on the own storage on the `addReserve` function

Reply:
then, you also do not need to check for the ethertoken here

Reply:
by the way: you're very often reading reserveTokens[0] and reserveTokens[1] on this method. it is each time reading from a mapping from storage, which costs 300 gas each time. it should be way cheaper to read them in the beginning and store them locally, afterwards reference the local variable.

Reply:
Ok, I will store the decimals when the tokens are added to converter

Reply:
And also do the calculations once